File: lib/citations.ts
import type { CitationStyle, TitleRow } from "./types";

function splitAuthor(author: string): { last: string; first: string } {
  const a = (author || "").trim();
  if (!a) return { last: "", first: "" };
  if (a.includes(",")) {
    const [last, ...rest] = a.split(",");
    return { last: last.trim(), first: rest.join(",").trim() };
  }
  const parts = a.split(/\s+/);
  if (parts.length === 1) return { last: parts[0], first: "" };
  return { last: parts[parts.length - 1], first: parts.slice(0, -1).join(" ") };
}

function initials(first: string): string {
  return (first.match(/[A-Za-z]+/g) || []).map((p) => p[0].toUpperCase() + ".").join(" ");
}

/** Format a single title as a bibliography entry in the chosen style. */
export function formatCitation(t: TitleRow, style: CitationStyle = "apa7"): string {
  const title = (t.title || "").trim().replace(/\.$/, "");
  const year = (t.year || "").trim() || "n.d.";
  const publisher = (t.publisher || "").trim().replace(/\.$/, "");
  const { last, first } = splitAuthor(t.author || "");

  if (style === "apa7") {
    const author = last ? `${last}, ${initials(first)}`.replace(/,\s*$/, "") : "";
    const head = author ? `${author} (${year}). ` : `(${year}). `;
    const tail = publisher ? `${title}. ${publisher}.` : `${title}.`;
    return head + tail;
  }
  if (style === "mla9") {
    const author = last ? `${last}, ${first}`.replace(/,\s*$/, "") : "";
    return `${author ? author + ". " : ""}${title}. ${publisher}, ${year}.`.trim();
  }
  if (style === "chicago") {
    const author = last ? `${last}, ${first}`.replace(/,\s*$/, "") : "";
    return `${author ? author + ". " : ""}${year}. ${title}. ${publisher}.`.trim();
  }
  // harvard
  const author = last ? `${last}, ${initials(first)}`.replace(/,\s*$/, "") : "";
  const head = author ? `${author} ${year}, ` : `${year}, `;
  return (`${head}${title}, ${publisher}`).replace(/,\s*$/, "") + ".";
}

export const CITATION_STYLES: { id: CitationStyle; label: string }[] = [
  { id: "apa7",    label: "APA 7" },
  { id: "mla9",    label: "MLA 9" },
  { id: "chicago", label: "Chicago" },
  { id: "harvard", label: "Harvard" },
];

File: app/citations.py
"""Citation formatting in APA 7, MLA 9, Chicago (author-date), and Harvard."""
from __future__ import annotations

import re

STYLES = ("apa7", "mla9", "chicago", "harvard")


def _split_author(author: str) -> tuple[str, str]:
    """Return (last, initials_or_first). Handles 'Jane Q. Doe' and 'Doe, Jane Q.'."""
    author = (author or "").strip()
    if not author:
        return "", ""
    if "," in author:
        last, rest = [p.strip() for p in author.split(",", 1)]
        return last, rest
    parts = author.split()
    if len(parts) == 1:
        return parts[0], ""
    return parts[-1], " ".join(parts[:-1])


def _first_initials(first: str) -> str:
    parts = re.findall(r"[A-Za-z]+", first)
    return " ".join(p[0].upper() + "." for p in parts)


def format_citation(title: dict, style: str = "apa7") -> str:
    style = (style or "apa7").lower()
    t = (title.get("title") or "").strip().rstrip(".")
    author = (title.get("author") or "").strip()
    year = (title.get("year") or "n.d.").strip() or "n.d."
    publisher = (title.get("publisher") or "").strip().rstrip(".")
    last, first = _split_author(author)

    if style == "apa7":
        author_str = f"{last}, {_first_initials(first)}".strip().rstrip(",") if last else ""
        head = f"{author_str} ({year}). " if author_str else f"({year}). "
        tail = f"{t}. {publisher}." if publisher else f"{t}."
        return head + tail
    if style == "mla9":
        author_str = f"{last}, {first}".strip().rstrip(",") if last else ""
        head = f"{author_str}. " if author_str else ""
        return f"{head}{t}. {publisher}, {year}.".strip()
    if style == "chicago":
        author_str = f"{last}, {first}".strip().rstrip(",") if last else ""
        head = f"{author_str}. " if author_str else ""
        return f"{head}{year}. {t}. {publisher}.".strip()
    if style == "harvard":
        author_str = f"{last}, {_first_initials(first)}".strip().rstrip(",") if last else ""
        head = f"{author_str} {year}, " if author_str else f"{year}, "
        return f"{head}{t}, {publisher}.".strip().rstrip(",") + "."
    return format_citation(title, "apa7")
